New FPGA code sample board_test (#2)
aktole/oneAPI-samples: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md

# Board Test

> This design contains tests to check FPGA board interfaces and reports the following metrics:

Reviewer: I'm sorry I'll have to make you move this test in "Reference Designs" as this is not demonstrating a tool.
```suggestion
This reference design contains tests to check FPGA board interfaces and reports the following metrics:

```

Author: updated readme

| Optimized for                     | Description
---                                 |---
| OS                                | Linux* Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE* 15
| Hardware                          | Intel® Programmable Acceleration Card (PAC) with Intel Arria® 10 GX FPGA <br> Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX) <br> Intel® FPGA 3rd party / custom platforms with oneAPI support <br> *__Note__: Intel® FPGA PAC hardware is only compatible with Ubuntu 18.04*
| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit
| What you will learn               | How to test board interfaces to ensure designed platform provides expected performance
| Time to complete                  | 30 minutes (not including compile time)

## Purpose
This test can be used to check FPGA board interfaces and measure host to device as well as kernel to global memory interface metrics. Custom platform developers can use this test as a starting point to validate custom platform interfaces.

### Introduction to oneAPI FPGA Board Support Package(BSP)

> A oneAPI BSP consists of software layers and an FPGA hardware scaffold design that makes it possible to target the FPGA through the Intel® oneAPI DPC++/C++ Compiler. The FPGA design generated by the compiler is stitched into the framework provided by the oneAPI BSP. Please see information about oneAPI BSPs in [oneAPI Programming Guide](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/programming-interface/fpga-flow/fpga-bsps-and-boards.html). 

Reviewer: ```suggestion
A oneAPI board support package (BSP) consists of software layers and an FPGA hardware scaffold design that makes it possible to target the FPGA through the Intel® oneAPI DPC++/C++ Compiler. The FPGA design generated by the compiler is stitched into the framework provided by the oneAPI BSP. Please see information about oneAPI BSPs in [oneAPI Programming Guide](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/programming-interface/fpga-flow/fpga-bsps-and-boards.html). 

```

The BSP hardware components typically comprise RTL for all interfaces required by the oneAPI kernel, for example a PCIe IP for host to kernel communication, EMIF (External Memory Interface) IP for kernel to memory as well as host to FPGA board memory communication etc.

> The BSP software components typically consists of a MMD (Memory Mapped Device) layer and a driver. The implementation is vendor dependent.

Reviewer: ```suggestion
The BSP software components typically consists of a Memory Mapped Device (MMD) layer and a driver. The implementation is vendor dependent.

```

*Note:* The test allows overriding this failure, however, this is may lead to functional errors and is not recommended. The override option is provided to allow debug in case where platform design changes are done to force kernel to run at slower clock (this is not a common use-case). Usage section describes how to override this option if required.

The complete test is divided into six sub-tests. By default, all tests are run. You can choose to run a single test using `-test=<test number>` option. Please see [Running the Sample](#running-the-sample) for test usage instructions.
1. Host Speed and Host Read Write Test
2. Kernel Clock Frequency Test
3. Kernel Launch Test
4. Kernel Latency Measurement
5. Kernel-to-Memory Read Write Test
6. Kernel-to-Memory Bandwidth Test<br>
*Note:* It is recommended to run all tests at least once to ensure that the platform interfaces are fully functional.

> ## Key Concepts

Reviewer: I think that you can remove this section as I feel it does not bring much information

Author: done

* Sample to demonstrate platform interface checks

## License  
Code samples are licensed under the MIT license. See
[License.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/License.txt) for details.

Third party program Licenses can be found here: [third-party-programs.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt)

## Building the `board_test` code sample

> 

Reviewer: ```suggestion

> **Note**: If you have not already done so, set up your CLI
> environment by sourcing  the `setvars` script located in
> the root of your oneAPI installation.
>
> Linux Sudo: . /opt/intel/oneapi/setvars.sh
>
> Linux User: . ~/intel/oneapi/setvars.sh
>
> Windows: C:\Program Files(x86)\Intel\oneAPI\setvars.bat
>
>For more information on environment variables, see Use the setvars Script for [Linux or macOS](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html), or [Windows](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html).
```

### Include Files
The included header `dpc_common.hpp` is located at `$ONEAPI_ROOT/dev-utilities/latest/include` on your development system.

### Running Samples in DevCloud
If running a sample in the Intel DevCloud, remember that you must specify the type of compute node and whether to run in batch or interactive mode. Compiles to FPGA are only supported on fpga_compile nodes. Executing programs on FPGA hardware is only supported on fpga_runtime nodes of the appropriate type, such as fpga_runtime:arria10 or fpga_runtime:stratix10.  Neither compiling nor executing programs on FPGA hardware are supported on the login nodes. For more information, see the [Intel® oneAPI Base Toolkit Get Started Guide](https://devcloud.intel.com/oneapi/documentation/base-toolkit/).

When compiling for FPGA hardware, it is recommended to increase the job timeout to 12h.

> 

Reviewer: If you are planning on adding Windows support, you'll have to add a section on how to compile on Windows

Author: Added a section on Windows, will be in next upload

   ```
   Alternatively, to compile for the Intel® PAC with Intel Arria® 10 GX FPGA, run `cmake` using the command:
   ```
   cmake -DFPGA_BOARD=intel_a10gx_pac:pac_a10 ..
   ```
   You can also compile for a custom FPGA platform. Ensure that the board support package is installed on your system. Then run `cmake` using the command:
   ```
   cmake -DFPGA_BOARD=<board-support-package>:<board-variant> ..
   ```

2. Compile the design through the generated `Makefile`. The following build targets are provided, matching the recommended development flow:

   * Compile for emulation (compiles quickly, targets emulated FPGA device):
      ```
      make fpga_emu
      ```
   * Generate the optimization report: 
     ```
     make report
     ```
   * Compile for FPGA hardware (takes longer to compile, targets FPGA device):
     ```
     make fpga
     ```
3. (Optional) As the above hardware compile may take several hours to complete, FPGA pre-compiled binaries (compatible with Linux* Ubuntu* 18.04) can be downloaded <a href="https://iotdk.intel.com/fpga-precompiled-binaries/latest/board_test.fpga.tar.gz" download>here</a>.

### In Third-Party Integrated Development Environments (IDEs)

> You can compile and run this tutorial in the Eclipse* IDE (in Linux*). For instructions, refer to the following link: [Intel® oneAPI DPC++ FPGA Workflows on Third-Party IDEs](https://software.intel.com/en-us/articles/intel-oneapi-dpcpp-fpga-workflow-on-ide)

Reviewer: ```suggestion
You can compile and run this tutorial in the Eclipse* IDE (in Linux*). For instructions, refer to the following link: [Intel® oneAPI DPC++ FPGA Workflows on Third-Party IDEs](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-oneapi-dpcpp-fpga-workflow-on-ide.html)

```

## Examining the Reports
Locate `report.html` in the `board_test.prj/reports/` directory. Open the report in any of Chrome*, Firefox*, Edge*, or Internet Explorer*.

You may also wish to examine the reports generated by the full FPGA hardware compile and compare their contents.

## Running the Sample

 1. Run the sample on the FPGA emulator (the kernel executes on the CPU):
     ```

>      ./board_test.fpga_emu     (Linux)

Reviewer: If you add Windows support, you'll need to add Windows commands as well

Author: I have added Windows section, it will be in next upload

     ```
2. Run the sample on the FPGA device:
     ```

>      ./board_test.fpga         (Linux)

Reviewer: If you add Windows support, you'll need to add Windows commands as well

Author: I have added Windows section, it will be in next upload

     ```
The above commands run all the test described in [Test Details](#test-details), to run a specific test pass the test number as argument to `-test=<test number>` option: 
```

> ./board_test.fpga -test=<test number>     (Linux)

Reviewer: If you add Windows support, you'll need to add Windows commands as well

Author: I have added Windows section, it will be in next upload

If the reports folder is not found, the test will return the measured kernel clock frequency and fail, none of the other tests will run as hardware frequency may not be the expected value. Continuing to run other tests may lead to functional errors.<br>
*Note:* The test allows overriding this failure, however, this is may lead to functional errors and is not recommended. The override option is provided to allow debug in case where platform design changes are done to force kernel to run at slower clock (this is not a common use-case). To override, set `report_chk` variable to `false` in `board_test.cpp` and recompile host code only using `-reuse-exe=board_test.fpga` option in compile command.

2. Requirement for kernel-to-memory bandwidth test: Export `HLD_SHIM_ROOT_HW` to point to `path-to-oneAPI-shim/hardware/<board_variant>` directory. The test looks for `board_spec.xml` file in this folder to get the global memory information for the board. The test does not fail if this variable is not set, it reports measured kernel to device global memory bandwidth numbers and continues. 

To view test details and usage information using the binary, use `-help` option.
```

> ./board_test.fpga -help      (Linux)

Reviewer: If you add Windows support, you'll need to add Windows commands as well

Author: I have added Windows section, it will be in next upload

*****************************************************************

Note: This test assumes that design was compiled with -Xsno-interleaving option

Size of memory bank 1 = 0x200000000 bytes 
Size of memory bank 2 = 0x200000000 bytes 
Size of memory bank 3 = 0x200000000 bytes 
Size of memory bank 4 = 0x200000000 bytes 

Performing kernel transfers of 4096 MBs on the default global memory (address starting at 0)
Launching kernel MemWriteStream ... 
Launching kernel MemReadStream ... 
Launching kernel MemReadWriteStream ... 

Summarizing bandwidth in MB/s/bank for banks 1 to 4
 15862.1  15874.6  15871.7  15879.5  MemWriteStream
 17339.5  17339.9  17339.5  17339.8  MemReadStream
 16014.4  16018.5  16018.2  16018.5  MemReadWriteStream

Name of the global memory type      :    DDR
Number Of Interfaces            :    4
Max Bandwidth (all memory interfaces)   :    76800 MB/s 
Max Bandwidth of 1 memory interface in board_spec.xml :    19200 MB/s 

It is assumed that all memory interfaces have equal widths. 

BOARD BANDWIDTH UTILIZATION = 85.4672
Warning : Board bandwidth utilization is less than 90% 

KERNEL-TO-MEMORY BANDWIDTH = 16409.7 MB/s/bank

BOARD TEST PASSED
 

> ```

Reviewer: We usually have an additional section for Reference Designs, e.g.:
```
## Additional Design Information
### Source Code Explanation

| File                         | Description
---                            |---
| `gzip.cpp`                   | Contains the `main()` function and the top-level interfaces to the SYCL* GZIP functions.
| `gzip_ll.cpp`                | Low latency variant of the top level file.
| `gzipkernel.cpp`             | Contains the SYCL* kernels used to implement GZIP.
| `gzipkernel_ll.cpp`          | Low-latency variant of kernels.
...

### Compiler Flags Used

| Flag | Description
---    |---
`-Xshardware` | Target FPGA hardware (as opposed to FPGA emulator)
`-Xsparallel=2` | Uses two cores when compiling the bitstream through Quartus®
`-Xsseed=<seed_num>` | Uses a particular seed while running Quartus®, selected to yield the best Fmax for this desgin
`-Xsnum-reorder=6` | On Intel Stratix® 10 SX only, specify a wider data path for read data from global memory
`-Xsopt-arg="-nocaching"` | Specifies that cached LSUs should not be used.
`-DNUM_ENGINES=<1|2>` | Specifies that 1 GZIP engine should be compiled when targeting Intel Arria® 10 GX and two engines when targeting Intel Stratix® 10 SX
```

Author: Added section with source file and compiler flag information
